**In this notebook we'll run some model tests and look at a base model to get an initial idea of our model summary and perform inital feature selection/filtering.**

In [1]:
import pandas as pd
import pickle as pkl
from module2_scripts.model import ModelTests, BaseModel

In [2]:
X = pd.read_csv('working_datasets/df_preprocessed_X.csv')
y = pd.read_csv('working_datasets/df_preprocessed_y.csv') 

In [3]:
#initialize ModelTests instance object
obj = ModelTests(X, y)

Note***
	Converted date object column 'date'values to integer number of days since start date
	Start date: 2014-05-02


In [4]:
#initialize .run_tests() method to display OLS reports for each variable
obj.run_tests()

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.142
Model:                            OLS   Adj. R-squared:                  0.142
Method:                 Least Squares   F-statistic:                     3354.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8223e+05
No. Observations:               20332   AIC:                         5.645e+05
Df Residuals:                   20330   BIC:                         5.645e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -3.58e+07   6.27e+05    -57.093      0.000    -3.7e+07   -3.46e+07
lat         7.634e+05   1.32e+04     57.911      0.000    7.38e+05    7.89e+05
==============================================================================
Omnibus:                     8258.070   Durbin-Watson:                   1.955
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            39466.798
Skew:                           1.950   Prob(JB):                         0.00
Kurtosis:                       8.602   Cond. No.                     1.65e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.980
Date:                Sat, 12 Dec 2020   Prob (F-statistic):              0.159
Time:                        23:36:27   Log-Likelihood:            -2.8378e+05
No. Observations:               20332   AIC:                         5.676e+05
Df Residuals:                   20330   BIC:                         5.676e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.978e+06   1.75e+06      1.700      0.089   -4.56e+05    6.41e+06
long        2.017e+04   1.43e+04      1.407      0.159   -7927.081    4.83e+04
==============================================================================
Omnibus:                     7326.685   Durbin-Watson:                   1.961
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27864.739
Skew:                           1.793   Prob(JB):                         0.00
Kurtosis:                       7.475   Cond. No.                     1.10e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.1e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.087
Method:                 Least Squares   F-statistic:                     1949.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8285e+05
No. Observations:               20332   AIC:                         5.657e+05
Df Residuals:                   20330   BIC:                         5.657e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         5.125e+05   1868.408    274.284      0.000    5.09e+05    5.16e+05
bedrooms_log  8.248e+04   1868.454     44.143      0.000    7.88e+04    8.61e+04
==============================================================================
Omnibus:                     6995.742   Durbin-Watson:                   1.959
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            26416.399
Skew:                           1.705   Prob(JB):                         0.00
Kurtosis:                       7.422   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.193
Model:                            OLS   Adj. R-squared:                  0.193
Method:                 Least Squares   F-statistic:                     4866.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8160e+05
No. Observations:               20332   AIC:                         5.632e+05
Df Residuals:                   20330   BIC:                         5.632e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          5.125e+05   1756.897    291.693      0.000    5.09e+05    5.16e+05
bathrooms_log  1.226e+05   1756.940     69.760      0.000    1.19e+05    1.26e+05
==============================================================================
Omnibus:                     6485.997   Durbin-Watson:                   1.956
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            23044.075
Skew:                           1.594   Prob(JB):                         0.00
Kurtosis:                       7.128   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.366
Model:                            OLS   Adj. R-squared:                  0.366
Method:                 Least Squares   F-statistic:                 1.175e+04
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.7914e+05
No. Observations:               20332   AIC:                         5.583e+05
Df Residuals:                   20330   BIC:                         5.583e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const            5.125e+05   1556.922    329.159      0.000    5.09e+05    5.16e+05
sqft_living_log  1.688e+05   1556.961    108.419      0.000    1.66e+05    1.72e+05
==============================================================================
Omnibus:                     5742.355   Durbin-Watson:                   1.975
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            19659.018
Skew:                           1.413   Prob(JB):                         0.00
Kurtosis:                       6.901   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.012
Model:                            OLS   Adj. R-squared:                  0.012
Method:                 Least Squares   F-statistic:                     241.5
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           3.88e-54
Time:                        23:36:27   Log-Likelihood:            -2.8366e+05
No. Observations:               20332   AIC:                         5.673e+05
Df Residuals:                   20330   BIC:                         5.673e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         5.125e+05   1944.388    263.566      0.000    5.09e+05    5.16e+05
sqft_lot_log  3.021e+04   1944.436     15.539      0.000    2.64e+04     3.4e+04
==============================================================================
Omnibus:                     6973.884   Durbin-Watson:                   1.949
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            25160.195
Skew:                           1.720   Prob(JB):                         0.00
Kurtosis:                       7.227   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.076
Model:                            OLS   Adj. R-squared:                  0.076
Method:                 Least Squares   F-statistic:                     1669.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8298e+05
No. Observations:               20332   AIC:                         5.660e+05
Df Residuals:                   20330   BIC:                         5.660e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.125e+05   1880.253    272.556      0.000    5.09e+05    5.16e+05
floors_log  7.681e+04   1880.299     40.851      0.000    7.31e+04    8.05e+04
==============================================================================
Omnibus:                     7277.580   Durbin-Watson:                   1.963
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            28507.276
Skew:                           1.766   Prob(JB):                         0.00
Kurtosis:                       7.601   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.408
Model:                            OLS   Adj. R-squared:                  0.408
Method:                 Least Squares   F-statistic:                 1.399e+04
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.7846e+05
No. Observations:               20332   AIC:                         5.569e+05
Df Residuals:                   20330   BIC:                         5.569e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.125e+05   1505.432    340.417      0.000     5.1e+05    5.15e+05
grade_log    1.78e+05   1505.469    118.267      0.000    1.75e+05    1.81e+05
==============================================================================
Omnibus:                     6117.327   Durbin-Watson:                   1.950
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            22381.102
Skew:                           1.484   Prob(JB):                         0.00
Kurtosis:                       7.196   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     18.98
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           1.33e-05
Time:                        23:36:27   Log-Likelihood:            -2.8377e+05
No. Observations:               20332   AIC:                         5.676e+05
Df Residuals:                   20330   BIC:                         5.676e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         5.125e+05   1954.988    262.137      0.000    5.09e+05    5.16e+05
yr_built_log  8517.4895   1955.036      4.357      0.000    4685.460    1.23e+04
==============================================================================
Omnibus:                     7321.253   Durbin-Watson:                   1.963
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27759.378
Skew:                           1.793   Prob(JB):                         0.00
Kurtosis:                       7.462   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     37.20
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           1.08e-09
Time:                        23:36:27   Log-Likelihood:            -2.8376e+05
No. Observations:               20332   AIC:                         5.675e+05
Df Residuals:                   20330   BIC:                         5.675e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          5.125e+05   1954.114    262.254      0.000    5.09e+05    5.16e+05
condition_log  1.192e+04   1954.162      6.099      0.000    8088.671    1.57e+04
==============================================================================
Omnibus:                     7295.821   Durbin-Watson:                   1.959
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27670.207
Skew:                           1.786   Prob(JB):                         0.00
Kurtosis:                       7.461   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.6061
Date:                Sat, 12 Dec 2020   Prob (F-statistic):              0.436
Time:                        23:36:27   Log-Likelihood:            -2.8378e+05
No. Observations:               20332   AIC:                         5.676e+05
Df Residuals:                   20330   BIC:                         5.676e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5.149e+05   3681.569    139.860      0.000    5.08e+05    5.22e+05
date_delta   -13.4729     17.305     -0.779      0.436     -47.393      20.447
==============================================================================
Omnibus:                     7313.350   Durbin-Watson:                   1.960
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27746.632
Skew:                           1.791   Prob(JB):                         0.00
Kurtosis:                       7.464   Cond. No.                         400.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.241
Model:                            OLS   Adj. R-squared:                  0.241
Method:                 Least Squares   F-statistic:                     808.2
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8098e+05
No. Observations:               20332   AIC:                         5.620e+05
Df Residuals:                   20323   BIC:                         5.620e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const              5.57e+05   2837.185    196.329      0.000    5.51e+05    5.63e+05
zipcode_dummy__B      3e+04   5483.313      5.470      0.000    1.92e+04    4.07e+04
zipcode_dummy__C  1.555e+05   5683.396     27.363      0.000    1.44e+05    1.67e+05
zipcode_dummy__D -2.488e+05   6749.180    -36.862      0.000   -2.62e+05   -2.36e+05
zipcode_dummy__E  -2.18e+05   7306.566    -29.842      0.000   -2.32e+05   -2.04e+05
zipcode_dummy__F  -2.74e+05   6722.902    -40.750      0.000   -2.87e+05   -2.61e+05
zipcode_dummy__G   5.95e+04   6648.045      8.950      0.000    4.65e+04    7.25e+04
zipcode_dummy__H -2.373e+05   7376.904    -32.173      0.000   -2.52e+05   -2.23e+05
zipcode_dummy__I -9.535e+04   1.11e+04     -8.593      0.000   -1.17e+05   -7.36e+04
==============================================================================
Omnibus:                     7790.698   Durbin-Watson:                   1.957
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            37816.682
Skew:                           1.814   Prob(JB):                         0.00
Kurtosis:                       8.610   Cond. No.                         7.06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     486.4
Date:                Sat, 12 Dec 2020   Prob (F-statistic):          1.53e-106
Time:                        23:36:27   Log-Likelihood:            -2.8354e+05
No. Observations:               20332   AIC:                         5.671e+05
Df Residuals:                   20330   BIC:                         5.671e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                5.096e+05   1937.302    263.046      0.000    5.06e+05    5.13e+05
waterfront_dummy__1  6.352e+05   2.88e+04     22.055      0.000    5.79e+05    6.92e+05
==============================================================================
Omnibus:                     7018.252   Durbin-Watson:                   1.954
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            25572.660
Skew:                           1.727   Prob(JB):                         0.00
Kurtosis:                       7.272   Cond. No.                         14.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.128
Model:                            OLS   Adj. R-squared:                  0.128
Method:                 Least Squares   F-statistic:                     744.7
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:27   Log-Likelihood:            -2.8239e+05
No. Observations:               20332   AIC:                         5.648e+05
Df Residuals:                   20327   BIC:                         5.648e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          4.835e+05   1913.799    252.662      0.000     4.8e+05    4.87e+05
view_dummy__1   2.49e+05    1.5e+04     16.611      0.000     2.2e+05    2.78e+05
view_dummy__2  2.424e+05   9142.227     26.510      0.000    2.24e+05     2.6e+05
view_dummy__3  3.968e+05   1.28e+04     30.910      0.000    3.72e+05    4.22e+05
view_dummy__4   6.04e+05   1.73e+04     34.952      0.000     5.7e+05    6.38e+05
==============================================================================
Omnibus:                     6545.624   Durbin-Watson:                   1.950
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            23432.284
Skew:                           1.607   Prob(JB):                         0.00
Kurtosis:                       7.164   Cond. No.                         9.48
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.033
Model:                            OLS   Adj. R-squared:                  0.032
Method:                 Least Squares   F-statistic:                     683.8
Date:                Sat, 12 Dec 2020   Prob (F-statistic):          2.86e-148
Time:                        23:36:27   Log-Likelihood:            -2.8345e+05
No. Observations:               20332   AIC:                         5.669e+05
Df Residuals:                   20330   BIC:                         5.669e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                   4.726e+05   2455.143    192.489      0.000    4.68e+05    4.77e+05
sqft_basement_dummy__1  1.033e+05   3951.731     26.149      0.000    9.56e+04    1.11e+05
==============================================================================
Omnibus:                     7299.109   Durbin-Watson:                   1.958
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27914.959
Skew:                           1.783   Prob(JB):                         0.00
Kurtosis:                       7.498   Cond. No.                         2.44
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

'--- Sum of Squared Errors Summary Report ---'

column                  sse
1         grade_log  936784301822452.000
2   sqft_living_log 1001962144904332.250
3           zipcode 1199623013751760.750
4     bathrooms_log 1275880394742519.000
5               lat 1357375203082427.750
6              view 1379189468704001.250
7      bedrooms_log 1442981461432323.250
8        floors_log 1461335292350598.000
9     sqft_basement 1529835318240923.250
10       waterfront 1544339263022662.000
11     sqft_lot_log 1562728249176744.250
12    condition_log 1578400197659754.500
13     yr_built_log 1579813498745357.000
14             long 1581134465531667.500
15       date_delta 1581241320410129.500

The above cell executed calls the run_tests() method on the MotelTests instance object and returns the OLS regresseion summary report for each variable, and calculated and displays the sum of squares for each model run, ordered from least to greatest. We generally want to keep the variables that have lower SSE. For categorical variables that have been one-hot encoded, the method groups all the dummy variables together and then runs the model, you can see the 'view' and 'zipcode' OLS summaries for reference. 

Individually, none of the variables have strong R-squared figures. But, we expect the overall base model R-squared to increase once all the variables are grouped together and then the model is ran, and then we expect R-squared to increase further once we do further feature selection and introduce interaction variables into the model. 

For now, we will focus on again filtering the features manually, considering which features would it make sense to leave in the model and which the model would be better off without, and also considering the SSE for each variable. 

To start, focusing our attention on the latitude and longitude variable pairs, these coordinates do not work very well with the model. These have not been log transformed as these are moreso descriptor variables rather than continuous variables. Also, the zipcode variable and the lat-long variable pairs more or less tell us similar info, which is info regarding the geographic location of the unit. Since zipcode ranked #3 in lowest SSE, it makes sense to keep the zipcode variables and drop the **lat and long** variable pairs. 

Another clear variable to drop is the **date_delta** variable. This is a derivative of the date variable. Since the statsmodels.api OLS function does not work with date objects, this variable had to be converted to the n number of days since the earliest date in the set. 


In [5]:
#initialize BaseModel object
obj_base = BaseModel(X, y)

In [6]:
obj.ols_summary()

---------- Running OLS Regression Summary Report for ModelTests Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.696
Model:                            OLS   Adj. R-squared:                  0.695
Method:                 Least Squares   F-statistic:                     1858.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:28   Log-Likelihood:            -2.7168e+05
No. Observations:               20332   AIC:                         5.434e+05
Df Residuals:                   20306   BIC:                         5.436e+05
Df Model:                          25                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                  -3.171e+07   2.67e+06    -11.872      0.000    -3.7e+07   -2.65e+07
lat                     2.204e+05   1.92e+04     11.508      0.000    1.83e+05    2.58e+05
long                   -1.779e+05    2.1e+04     -8.468      0.000   -2.19e+05   -1.37e+05
bedrooms_log           -1.725e+04   1476.318    -11.682      0.000   -2.01e+04   -1.44e+04
bathrooms_log           1.361e+04   1934.280      7.038      0.000    9822.628    1.74e+04
sqft_living_log         9.359e+04   2412.393     38.794      0.000    8.89e+04    9.83e+04
sqft_lot_log            1930.8426   1586.345      1.217      0.224   -1178.521    5040.207
floors_log              1.443e+04   1697.049      8.503      0.000    1.11e+04    1.78e+04
grade_log               1.098e+05   1780.248     61.660      0.000    1.06e+05    1.13e+05
yr_built_log           -6.293e+04   1696.880    -37.085      0.000   -6.63e+04   -5.96e+04
condition_log           1.108e+04   1193.016      9.289      0.000    8743.794    1.34e+04
waterfront_dummy__1     2.785e+05   1.92e+04     14.523      0.000    2.41e+05    3.16e+05
view_dummy__1           9.861e+04   8968.710     10.995      0.000     8.1e+04    1.16e+05
view_dummy__2           8.271e+04   5546.801     14.911      0.000    7.18e+04    9.36e+04
view_dummy__3            1.65e+05   7801.655     21.151      0.000     1.5e+05     1.8e+05
view_dummy__4           2.776e+05   1.22e+04     22.660      0.000    2.54e+05    3.02e+05
sqft_basement_dummy__1 -1.228e+04   2857.354     -4.299      0.000   -1.79e+04   -6684.318
zipcode_dummy__B        4590.8532   5597.424      0.820      0.412   -6380.550    1.56e+04
zipcode_dummy__C        1.072e+05   5646.699     18.983      0.000    9.61e+04    1.18e+05
zipcode_dummy__D       -1.334e+05   6774.347    -19.698      0.000   -1.47e+05    -1.2e+05
zipcode_dummy__E       -1.246e+05   6314.777    -19.732      0.000   -1.37e+05   -1.12e+05
zipcode_dummy__F       -1.572e+05   8217.111    -19.127      0.000   -1.73e+05   -1.41e+05
zipcode_dummy__G         2.55e+04   8060.881      3.164      0.002    9704.858    4.13e+04
zipcode_dummy__H       -7.785e+04   9475.365     -8.216      0.000   -9.64e+04   -5.93e+04
zipcode_dummy__I        2.118e+04   1.28e+04      1.654      0.098   -3916.720    4.63e+04
date_delta                96.1844      9.590     10.030      0.000      77.388     114.981
==============================================================================
Omnibus:                     7208.950   Durbin-Watson:                   1.957
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            43275.972
Skew:                           1.582   Prob(JB):                         0.00
Kurtosis:                       9.409   Cond. No.                     6.00e+05
=================================================

Running the OLS regression for the base model, we see that adjusted R-squared is 69.5%! Overall, the tests are okay given the nature of the dataset and the fact that the dependent variable has not been log transformed, however, the report gives us a warning that the condition number is large, which might indicate strong multicollinearity or "other numerical problems". Let's drop the aforementioned variables and rerun the base model and note the changes. 

In [7]:
X_processed = X.drop(['lat', 'long', 'date_delta'], axis=1)

In [8]:
obj_base_processed = BaseModel(X_processed, y)
obj_base_processed.ols_summary()

---------- Running OLS Regression Summary Report for BaseModel Object----------


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.691
Model:                            OLS   Adj. R-squared:                  0.691
Method:                 Least Squares   F-statistic:                     2065.
Date:                Sat, 12 Dec 2020   Prob (F-statistic):               0.00
Time:                        23:36:28   Log-Likelihood:            -2.7184e+05
No. Observations:               20332   AIC:                         5.437e+05
Df Residuals:                   20309   BIC:                         5.439e+05
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                   5.618e+05   2771.568    202.713      0.000    5.56e+05    5.67e+05
bedrooms_log           -1.645e+04   1486.954    -11.064      0.000   -1.94e+04   -1.35e+04
bathrooms_log           1.303e+04   1948.805      6.685      0.000    9208.299    1.68e+04
sqft_living_log         9.334e+04   2429.850     38.415      0.000    8.86e+04    9.81e+04
sqft_lot_log            1618.6747   1579.654      1.025      0.306   -1477.574    4714.923
floors_log              1.288e+04   1707.614      7.545      0.000    9537.718    1.62e+04
grade_log                1.11e+05   1788.380     62.047      0.000    1.07e+05    1.14e+05
yr_built_log           -6.367e+04   1704.833    -37.344      0.000    -6.7e+04   -6.03e+04
condition_log           1.034e+04   1199.901      8.613      0.000    7983.407    1.27e+04
waterfront_dummy__1     2.828e+05   1.93e+04     14.646      0.000    2.45e+05    3.21e+05
view_dummy__1           9.958e+04   9018.243     11.042      0.000    8.19e+04    1.17e+05
view_dummy__2           8.261e+04   5576.256     14.814      0.000    7.17e+04    9.35e+04
view_dummy__3           1.642e+05   7850.450     20.910      0.000    1.49e+05     1.8e+05
view_dummy__4           2.809e+05   1.23e+04     22.763      0.000    2.57e+05    3.05e+05
sqft_basement_dummy__1 -1.156e+04   2871.127     -4.027      0.000   -1.72e+04   -5934.343
zipcode_dummy__B       -1.299e+04   4237.228     -3.066      0.002   -2.13e+04   -4684.312
zipcode_dummy__C        5.774e+04   4285.385     13.474      0.000    4.93e+04    6.61e+04
zipcode_dummy__D       -2.036e+05   4764.644    -42.739      0.000   -2.13e+05   -1.94e+05
zipcode_dummy__E       -1.669e+05   5062.741    -32.959      0.000   -1.77e+05   -1.57e+05
zipcode_dummy__F       -2.396e+05   4885.165    -49.047      0.000   -2.49e+05    -2.3e+05
zipcode_dummy__G       -4.157e+04   4933.260     -8.426      0.000   -5.12e+04   -3.19e+04
zipcode_dummy__H        -1.91e+05   5430.990    -35.160      0.000   -2.02e+05    -1.8e+05
zipcode_dummy__I       -9.651e+04   7606.224    -12.689      0.000   -1.11e+05   -8.16e+04
==============================================================================
Omnibus:                     7346.823   Durbin-Watson:                   1.954
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            45298.272
Skew:                           1.608   Prob(JB):                         0.00
Kurtosis:                       9.567   Cond. No.                         35.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We see that our updated OLS report reflects a very minor decrease in adjusted R-squared to 69.1%. A decrease is expected since we removed some variables from the model. The more variables that are included in the model, the more R-squared artificially increases. The good news is that we dropped 3 variables and the R-squared only re-adjusted -0.4%, which shows that the remaining variables still hold a lot of value to the model. We also note that the condition number is very low and the multicollinearity and other numberical problems warning is now gone. 

Now, we're ready to proceed to the next step of the model, which is to do further feature selection using stepwise selection method, and add interaction variables for better model fit. 

We'll save the dependent variable and updated predictor variable datasets to be used in a different notebook.

In [9]:
X_processed.to_csv('working_datasets/X_processed.csv', index=False)
y.to_csv('working_datasets/y_processed.csv', index=False)